lib/vsc/mympirun/pmi/slurm.py

```diff
@@ -0,0 +1,260 @@
+            self.log.error("Total number of tasks is not equal ranks per node %s time number of nodes: %s",
+                           job_info['nranks'], dbgtxt)
+
+        if 'mem_per_node' in se:
+            job_info['nmem'] = se['mem_per_node']
+        elif 'mem_per_cpu' in se:
+            job_info['nmem'] = se['mem_per_cpu'] * se['cpus_on_node']
+        else:
+            self.log.debug("No memory specification found")
+
+        if 'job_gpus' in se:
+            try:
+                # this should fail when slurm switched to compressed repr (eg 0-3 instead of current 0,1,2,3)
+                ngpus = len(map(int, se['job_gpus'].split(',')))
+                job_info['ngpus'] = ngpus
+            except Exception as e:
+                self.log.error("Failed to get the number of gpus per node from %s: %s", se['job_gpus'], e)
+
+        return job_info
+
+    def _pmicmd_size_args(self, mpi_info):
+        """
+        Convert mpi_info into launcher list of args
+        """
+
+        # patterns to cleanup
+        cleanup = ['MEM', 'CPU', 'GPU', 'TASK', 'NODE', 'PROC']
+        cleanupre = re.compile(r'^SLURM_.*(' + '|'.join(cleanup) + ').*')
+        # whitelist, precedes cleanup
+        keep = ['NODELIST']
+        keepre = re.compile(r'^SLURM_.*(' + '|'.join(keep) + ').*')
+
+        removed = []
+        for k, v in os.environ.items():
+            if keepre.search(k):
+                continue
+            elif cleanupre.search(k):
+                removed.append("%s=%s" % (k, v))
+                del os.environ[k]
+
+        if removed:
+            self.log.debug("Unset environment variables %s", " ".join(removed))
+        else:
+            self.log.debug("No environment variables unset")
+
+        args = []
+
+        tn = mpi_info['tnodes']
+        # all available nodes
+        args.append("--nodes=%s" % tn)
+
+        # 1 rank per task
+        # tasks per node
+        tpn = mpi_info['nranks']
+        args.append("--ntasks=%s" % (tn * tpn))
+
+        # cores per task
+        cpt = mpi_info['ncores'] // tpn
+        # sanity check
+        if cpt * tpn != mpi_info['ncores']:
+            self.log.error("Imbalanced tasks per node %s vs cores per node %s; using %s cores per task",
+                           tpn, mpi_info['ncores'], cpt)
+
+        args.append("--cpus-per-task=%s" % cpt)
+
+        # redistribute all the memory to the tasks/cores
+        if mpi_info['nmem'] is None:
+            self.log.debug("No nmem specified found (assuming slurm.conf defaults)")
+        else:
+            args.append("--mem-per-cpu=%s" % (mpi_info['nmem'] // mpi_info['ncores']))
+
+        if mpi_info['ngpus'] is not None:
+            # gpus per task
+            gpt = mpi_info['ngpus'] // tpn
+            # sanity check
+            if gpt * tpn != mpi_info['ngpus']:
+                self.log.error("Imbalanced tasks per node %s vs gpus per node %s; using %s gpus per task",
```

Comment:
Then that would be a warning, no? You are fixing it, so it is not technically an error.